msprime/simulations.py

```diff
@@ -1471,6 +1489,198 @@ def print_history(self, output=sys.stdout):
+        """
+        Private function is used for doing all the calculation for
+        coalescence_rate_trajectory.
+        """
+        num_pops = self.num_populations
+        P = np.zeros([num_pops**2, num_pops**2])
+        IA = np.array(range(num_pops**2)).reshape([num_pops, num_pops])
+        Identity = np.eye(num_pops)
+        for x in range(num_pops):
+            for y in range(num_pops):
+                K_x = num_samples[x]
+                K_y = num_samples[y]
+                P[IA[x, y], IA[x, y]] = K_x * (K_y - (x == y))
+        P = P / np.sum(P)
+        epoch_breaks = [e.start_time for e in self.epochs]
+        steps_b = sorted(list(set(list(steps) + epoch_breaks)))
+        num_steps = len(steps_b)
+        r = np.zeros(num_steps)
+        p_t = np.zeros(num_steps)
+        mass_migration_objects = []
+        mass_migration_times = []
+        for demo in self.demographic_events:
+            if type(demo) == MassMigration:
+                mass_migration_objects.append(demo)
+                mass_migration_times.append(demo.time)
+        prev = 0
+        for j, time in enumerate(steps_b):
+            dt = time - prev
+            prev = time
+            N, M = self.pop_size_and_migration_at_t(time)
+            C = np.zeros([num_pops**2, num_pops**2])
+            for idx in range(num_pops):
+                C[IA[idx, idx], IA[idx, idx]] = 1 / (2 * max(min_pop_size, N[idx]))
+            dM = np.diag([sum(s) for s in M])
+            G = (np.kron(M - dM, Identity) + np.kron(Identity, M - dM)) - C
+            if time in mass_migration_times:
+                idx = mass_migration_times.index(time)
+                a = mass_migration_objects[idx].source
+                b = mass_migration_objects[idx].dest
+                p = mass_migration_objects[idx].proportion
+                S = np.eye(num_pops**2, num_pops**2)
+                for x in range(num_pops):
+                    if x == a:
+                        S[IA[a, a], IA[a, b]] = S[IA[a, a], IA[b, a]] = p * (1 - p)
+                        S[IA[a, a], IA[b, b]] = p ** 2
+                        S[IA[a, a], IA[a, a]] = (1 - p) ** 2
+                    else:
+                        S[IA[x, a], IA[x, b]] = S[IA[a, x], IA[b, x]] = p
+                        S[IA[x, a], IA[x, a]] = S[IA[a, x], IA[a, x]] = 1 - p
+                P = np.matmul(P, S)
+            P = np.matmul(P, _matrix_exponential(dt * G))
+            p_t[j] = np.sum(P)
+            r[j] = np.sum(np.matmul(P, C)) / np.sum(P)
+        idx_counter = 0
+        diff_mask = []
+        for i in range(len(steps)):
+            if(steps[i] != steps_b[i+idx_counter]):
+                diff_mask.append(i+idx_counter)
+                idx_counter += 1
+        r = np.delete(r, diff_mask)
+        p_t = np.delete(p_t, diff_mask)
+        return r, p_t
+
+    def pop_size_and_migration_at_t(self, t):
+        """
+        Given a time, t (in generations), return a tuple (N, M) where
```

Comment:
How about
"
Returns a tuple (N, M) of population sizes (N) and migration rates (M) at time ``t`` ago.

:param float t: The time ago.

:return: A tuple of arrays, of the same form as the population sizes and migration rate arrays of the demographic model.